README.md
## CloudWatchLogs client
### Description
Allows running cloudwatch insight queries from within vscode.
Removes the hussle of switching accounts when queries differante environments logs.

### Syntax
each query should looks like the following:
```
env;regions;log-group-name;duration[;max-results]
fields @timestamp, Level, MessageTemplate
| sort @timestamp desc
| limit 10
```

- `env`: name of environment to query, should usually correspond with [aws shared credentials profile][1] that will be used to run the query
- `regions`: the [regions][3] to run the query in (comma seperated). eg. 'us-east-1' or 'us-east-1,use-east-2' (without the quotes). 
- `log-group-name`: the [log group][2] to query, you can specify multiple names with "," as seperator or using wildcards ("*")
- `duration`: the time range in duration human readable syntax, eg. 1h, 30m, 2d etc.. Its also support explicit time range with "->". For example, 2020-09-03T01:40+00:00->2020-09-03T01:50+00:00 (omit the +00 to use local time)
- `max-results`: optional, the maximum results to fatch from the server (defaults to 1000, requires the setting seperator to be ';')

~~~
Since version 0.4 the setting seperator chnaged from ':' to ';' (to support the optional 'max-results')
~~~

~~~
'#' at the beginning of the line mark the line as ignored
'#@' at the beginning of a line will set the results tab's title
~~~


### Usage
just make sure your curser is somewhere inside a query and press `ctrl`+`enter` or run vscode command `Execute CloudWatchlogs query`
![simpleQuery][simpleQuery]

you can also highlight any part of existing query (as long as it contains the first line of args as mentions before)
![selectedQuery][selectedQuery]

click the magnifying glass button to open the log record in new page.  
this will also format json values
![openLogRecord][openLogRecord]

seperate vscode container view allows you to save query files for easy access.  
right click query file to open in new column
![contianerView][contianerView]

results view allows you to edit the query before refreshing the results
![editQuery][editQuery]

### AWS Authentication
CloudWatchLogs client will use the `env` parameter as the [aws shared credentials profile][1] to use.  
when no profile exists by that name, CloudWatchLogs client will fallback to [AWS environment variables][4]  

CloudWatchLogs client will store the credentials in memmory until they expire and only then will try to  
require new ones.

required permission for the active credentials are:
- logs:DescribeLogGroups
- logs:StartQuery
- logs:GetQueryResults

if you need to run some executable in order to prepare the share credentials profile, you can configure 
authentication command in the extension setting. The authentication command will be the first step each time
the client tries to regain the credentials.
![settings][settings]


[1]: https://docs.aws.amazon.com/ses/latest/DeveloperGuide/create-shared-credentials-file.html
[2]: https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/Working-with-log-groups-and-streams.html
[3]: https://docs.aws.amazon.com/general/latest/gr/cwl_region.html
[4]: https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-envvars.html#envvars-list

[simpleQuery]: https://github.com/dorki/vscode.cloudWatchLogs/blob/master/media/gifs/simpleQuery.gif?raw=true
[selectedQuery]: https://github.com/dorki/vscode.cloudWatchLogs/blob/master/media/gifs/selectedQuery.gif?raw=true 
[openLogRecord]: https://github.com/dorki/vscode.cloudWatchLogs/blob/master/media/gifs/openLogRecord.gif?raw=true 
[settings]: https://github.com/dorki/vscode.cloudWatchLogs/blob/master/media/gifs/settings.png?raw=true
[contianerView]: https://github.com/dorki/vscode.cloudWatchLogs/blob/master/media/gifs/contianerView.gif?raw=true
[editQuery]: https://github.com/dorki/vscode.cloudWatchLogs/blob/master/media/gifs/editQuery.gif?raw=true

CHANGELOG.md
### version 0.6.0
- move to aws-sdk v3 to support pagination

### version 0.5.5
- support log-groups pagination (thanks to @mbraude-chwy)

### version 0.5.4
- support aws partitions (us-gov, cn)

### version 0.5.3
- fix - paste events in the raw query editor

### version 0.5.2
- fix - log group names were not displayed correctly in the results view
- fix - escaping '<' and '>' in the query text (for using regex)

### version 0.5.1
- support predefind tab title (use #@TITLE as any line in the query text)

### version 0.5.0
- support multiple regions at once (comma sepeated)
- fix - 'duplicate' button triggered the "set title" button
- fix - "limit" keyword casued rendering of partial results

### version 0.4.6
- support custom tab title
- tab icon now indicate query status (completed, working, faulted/cancelled)

### version 0.4.5
- skipped

### version 0.4.4
- fix - rendering column was bas when new fields added during results fetching
- fix - column toggles ui

### version 0.4.3
- new command: GetHistory. Opens new tab with recent query history

### version 0.4.2
- paste plain text only in the query editor (in results view)
- add "duplicate" button to results view
- fix - "limit" keyword didnt work
- fix - rendering results

### version 0.4.1
- refresh results tab title while query is running
- fix - query cancellation

### version 0.4.0
- change the way results are sent to display, improving performance and responsiveness
- improve "open raw" in results, supporting table filter, columns visibilty and ordering
- support setting the max results, it can now be set to more then 1000 (see in the README)

### version 0.3.6
- update datatables package (security patch)
- remove "empty" folder in query files
- avoid excessive querying while waiting for query results from cloudwatch
- fix - refresh bug in "open raw" in results page

### version 0.3.5
- add table export to the results window

### version 0.3.4
- fix - query time range parsing

### version 0.3.3
- support time range
- extract results table column names from results and not from query
- refresh results view by pressing Ctrl+Enter in the raw query edit box

### version 0.3.2
- support custom query files locations (can be added in extension's setting)
- fix - results view - button's column should not be sortable
- fix - query files should not open in perview mode

### version 0.3.1
- show authentication progress
- support for field aliases (`as`) both in `fields` and in `display` commands
- added action buttons to log record view (`copy to clipboard` and `show raw json`)
- fix - results view opens too soon (in case of an error it used to be left blank)
- fix - supports for neglecting the "fields" command
- fix - closing results view while query in progress throw excpetion

### version 0.3.0
- added new ConatinerView for storing query files
- allow editing the query in the results view
- added context menu to the editor so right click can also fire the query
- fix authentication logic

### version 0.2.0
- support multiple logGroups 
- support logGroup name wildcard ('*')
- fixed datetime locale issues

### version 0.1.0
- first release :)

src/extension.ts
import * as asTable from 'as-table';
import { CloudWatchLogsClient, CloudWatchLogsServiceException } from '@aws-sdk/client-cloudwatch-logs';
import { LogGroup, GetLogRecordCommand, DescribeLogGroupsCommand, StartQueryCommand, GetQueryResultsCommand, ResultField, GetQueryResultsResponse, DescribeQueriesCommand } from '@aws-sdk/client-cloudwatch-logs';
import { paginateDescribeLogGroups } from '@aws-sdk/client-cloudwatch-logs';
import * as clipboardy from 'clipboardy';
import * as _ from 'lodash';
import * as vscode from 'vscode';
import { getEnvCredentials } from './authenticator';
import { BuildLogRecordHtml, BuildQueryResultsHtml, formatTime } from './htmlHelper';
import { parseQuery, Query } from './query';
import { Initialize as InitializeQueryFiles } from './queryFiles';
import { getFocusedTextSection } from './windowUtils';

export function activate(context: vscode.ExtensionContext) {

	InitializeQueryFiles(context);

	async function GoToLog(logPtr: string, env: string, region: string) {

		if (_.isEmpty(logPtr)) {
			vscode.window.showInformationMessage("Stats queries does not have log record pointer");
			return;
		}

		const credentials = await getEnvCredentials(env, region);
		const logsClient = new CloudWatchLogsClient({ credentials, region });

		const logRecordResponse = await logsClient.send(new GetLogRecordCommand({ logRecordPointer: logPtr }));
		const panel =
			vscode.window.createWebviewPanel(
				`Log${Date.now()}`,
				`Log`,
				vscode.ViewColumn.Active,
				{
					enableFindWidget: true,
					enableScripts: true,
					retainContextWhenHidden: true
				}
			);

		panel.webview.onDidReceiveMessage(
			async message => {
				const logJson =
					JSON.stringify(
						logRecordResponse.logRecord!,
						null,
						4);
				switch (message.command) {
					case 'copy':
						clipboardy.writeSync(logJson);
						vscode.window.showInformationMessage("log copied to clipboard");
						return;
					case 'openRaw':
						await vscode.window.showTextDocument(
							await vscode.workspace.openTextDocument({
								content: logJson,
								language: "json"
							}),
							vscode.ViewColumn.Active);
						return;
				}
			},
			undefined,
			context.subscriptions
		);

		panel.webview.html = BuildLogRecordHtml(logRecordResponse.logRecord!);
	}

	function formatResultsRawTable(fieldNames: string[], rows: string[][]): string {
		const table = [];

		for (const row of rows) {

			const txtTableRow: { [id: string]: string } = {};

			for (const [fieldName, value] of _.zip(fieldNames, row)) {
				if (value != undefined) {
					txtTableRow[fieldName!] = value;
				}
			}

			table.push(txtTableRow);
		}

		return asTable.configure({ delimiter: " | " })(table);
	}

	function createQueryWebViewPanel(query: Query): vscode.WebviewPanel {
		const panel =
			vscode.window.createWebviewPanel(
				`Query${Date.now()}`,
				`Results ${query.env} (0)`,
				{
					viewColumn: vscode.ViewColumn.Active,
					preserveFocus: true
				},
				{
					enableFindWidget: true,
					enableScripts: true,
					retainContextWhenHidden: true
				},
			);

		panel.webview.onDidReceiveMessage(
			async message => {
				switch (message.command) {
					case 'goToLog': {
						await GoToLog(message.text, query.env, message.region);
						return;
					}
					case 'openRaw': {
						await vscode.window.showTextDocument(
							await vscode.workspace.openTextDocument({
								content: formatResultsRawTable(message.fieldNames, message.rows),
								language: "json"
							}),
							vscode.ViewColumn.Active);
						return;
					}
					case 'refresh': {
						query.canceled = true;
						query = parseQuery(message.query, query);
						await executeQuery(query, panel);
						return;
					}
					case 'duplicate': {
						const duplicatedPanel = createQueryWebViewPanel(query);
						duplicatedPanel.webview.html =
							BuildQueryResultsHtml(
								context.extensionPath,
								query,
								{});
						return;
					}
					case 'changeTitle': {
						const title =
							await vscode.window.showInputBox({
								placeHolder: 'Set title text',
								prompt: 'Tab Title',
								value: query.title
							});

						if (title == undefined) {
							return;
						}
						else if (query.title) {
							panel.title = panel.title.replace(query.title, title);
						}
						else {
							panel.title = panel.title.replace("Results", title);
						}

						query.title = title;
						return;
					}
				}
			},
			undefined,
			context.subscriptions
		);

		panel.onDidDispose(
			() => query.canceled = true,
			null,
			context.subscriptions
		);

		return panel;
	}

	function unorderedEquals(values: string[], otherValues: string[]) {
		if (values.length !== otherValues.length) return false;
		if (_.difference(values, otherValues).length !== 0) return false;
		if (_.difference(otherValues, values).length !== 0) return false;
		return true;
	}

	const logGroupsCache: { [envRegion: string]: LogGroup[] } = {};

	async function executeQuery(query: Query, panel?: vscode.WebviewPanel) {
		const creds = await getEnvCredentials(query.env, query.regions[0]);

		const regionToLogsClientMap =
			_(query.regions).
				keyBy().
				mapValues(region => new CloudWatchLogsClient({ credentials: creds, region })).
				value();

		const logGroups = query.logGroup.split(',');
		const regionToLogGroupsMap: { [region: string]: string[] } = {};

		for (const [region, logsClient] of _.entries(regionToLogsClientMap)) {

			const logGroupsCacheKey = `${query.env}.${region}`;

			if (!logGroupsCache[logGroupsCacheKey]) {
				await vscode.window.withProgress({
					location: vscode.ProgressLocation.Notification,
					cancellable: false,
					title: `Fetching log groups from ${region} in '${query.env}''...`
				}, async () => {
					const regionLogGroups: LogGroup[] = [];
					for await (const page of paginateDescribeLogGroups({ client: logsClient }, {})) {
						regionLogGroups.push.apply(
							regionLogGroups,
							page.logGroups as LogGroup[]);
					}

					logGroupsCache[logGroupsCacheKey] = regionLogGroups;
				});
			}

			regionToLogGroupsMap[region] =
				_(logGroups).
					map(logGroup => new RegExp(`^${logGroup.replace(/\*/g, '.*')}$`, "i")).
					flatMap(
						logGroupRegex =>
							_.filter(
								logGroupsCache[logGroupsCacheKey],
								logGroup => logGroupRegex.test(logGroup.logGroupName ?? ""))).
					map(logGroup => logGroup.logGroupName!).
					uniq().
					value();
		}

		const regionToStartQueryIdMap: { [x: string]: string } = {};
		for (const [region, logsClient] of _.entries(regionToLogsClientMap)) {
			try {
				const startQueryResponse =
					await logsClient.send(
						new StartQueryCommand({
							startTime: query.times.start / 1000,
							endTime: query.times.end / 1000,
							queryString: query.query,
							logGroupNames: regionToLogGroupsMap[region],
							limit: query.maxResults
						}));

				regionToStartQueryIdMap[region] = startQueryResponse.queryId!;
			}
			catch (error) {
				const serviceException = error as CloudWatchLogsServiceException
				vscode.window.showErrorMessage(`${serviceException.name}, error: ${serviceException.message}`);
				return;
			}
		}

		const currentPanel = panel ?? createQueryWebViewPanel(query);

		await vscode.window.withProgress({
			location: vscode.ProgressLocation.Notification,
			cancellable: true,
			title: 'Loading...'
		}, async (progress, token) => {

			progress.report({ increment: 0 });

			currentPanel.iconPath = getPanelIconPath("hourglass");

			// build results page
			currentPanel.webview.html =
				BuildQueryResultsHtml(
					context.extensionPath,
					query,
					regionToLogGroupsMap);

			let results: ResultField[][] = [];
			const regionToQueryResultsMap: { [x: string]: GetQueryResultsResponse } = {};
			do {
				for (const [region, logsClient] of _.entries(regionToLogsClientMap)) {
					if (regionToQueryResultsMap[region]?.status !== "Complete") {
						regionToQueryResultsMap[region] = await logsClient.send(new GetQueryResultsCommand({ queryId: regionToStartQueryIdMap[region] }));
					}
				}

				if (query.canceled || token.isCancellationRequested) {
					break;
				}

				progress.report({ increment: 40 });

				if (_.some(regionToQueryResultsMap, queryResultsResponse => queryResultsResponse.results != undefined)) {

					results =
						_.flatMap(
							regionToQueryResultsMap,
							queryResultsResponse => queryResultsResponse.results ?? []);

					currentPanel.title = getPanelTitle(query, results.length);

					const fields =
						_(results).
							flatMap(queryResult => queryResult.map(_ => _.field!)).
							uniq().
							without("@ptr").
							value();

					if (fields.length > 0) {
						const newFieldExists = !unorderedEquals(fields, query.queryResultsFieldNames ?? []);
						if (newFieldExists) {
							query.queryResultsFieldNames = fields;
						}

						currentPanel.webview.postMessage({
							command: 'results',
							fieldNames: query.queryResultsFieldNames,
							fieldRefresh: newFieldExists,
							regionToQueryResultsMap: regionToQueryResultsMap
						});
					}
				}

				await new Promise(resolve => setTimeout(resolve, 1000));
			}
			while (
				_.some(
					regionToQueryResultsMap,
					queryResultsResponse => queryResultsResponse.status !== "Complete") &&
				!query.canceled &&
				!token.isCancellationRequested);

			currentPanel.title = getPanelTitle(query, results.length);
			currentPanel.iconPath =
				getPanelIconPath(
					query.canceled || token.isCancellationRequested
						? "error"
						: "done");
		});
	}

	function getPanelIconPath(status: string) {
		return vscode.Uri.joinPath(context.extensionUri, "media/panel", `${status}.svg`);
	}

	function getPanelTitle(query: Query, results: number) {
		return `${query.title ?? 'Results'} ${query.env} (${results})`;
	}

	async function getHistory(query: Query) {

		const env =
			!_.isEmpty(query.env)
				? query.env
				: await vscode.window.showInputBox({ placeHolder: 'Enter environment name' });
		if (env == undefined) {
			return;
		}

		const region =
			!_.isEmpty(query.regions)
				? query.regions[0]
				: await vscode.window.showInputBox({ placeHolder: 'Enter region' });
		if (region == undefined) {
			return;
		}

		const creds = await getEnvCredentials(env, region);
		const logsClient = new CloudWatchLogsClient({ credentials: creds, region });
		const describeQueriesResponse = await logsClient.send(new DescribeQueriesCommand({}));

		const queries =
			_(describeQueriesResponse.queries).
				map(
					query => {
						return {
							createTime: query.createTime,
							queryString: query.queryString?.substr(query.queryString?.indexOf("|") + 2)
						};
					}).
				groupBy(query => query.queryString).
				map(
					(queries, queryString) => {
						return {
							queryString,
							createTime: _(queries).map(query => query.createTime).max()!
						};
					}).
				orderBy(_ => _.createTime, "desc").
				value();

		let content = `Query History (${queries.length})\n\n`;
		for (const query of queries) {
			content += formatTime(query.createTime) + "\n";
			content += query.queryString.trim() + "\n";
			content += "\n";
		}

		await vscode.window.showTextDocument(
			await vscode.workspace.openTextDocument({ content }),
			vscode.ViewColumn.Active);
	}

	context.subscriptions.push(
		vscode.commands.registerCommand(
			'extension.runQuery',
			async () => {
				const query = parseQuery(getFocusedTextSection());
				await executeQuery(query);
			}));

	context.subscriptions.push(
		vscode.commands.registerCommand(
			'extension.getHistory',
			async () => {
				const query = parseQuery(getFocusedTextSection());
				await getHistory(query);
			}));
}